hrjobcontract/js/test/controllers/modal/modal-contract_test.js

```diff
@@ -0,0 +1,146 @@
+define([
+  'common/angular',
+  'common/angularMocks',
+  'job-contract/app'
+], function(angular) {
+  'use strict';
+
+  describe('ModalContractCtrl', function() {
+    var ctrl, $rootScope, $controller, $scope, $q, $uibModal, $httpBackend, $uibModalInstanceSpy,
+      $uibModalSpy, entity, ContractDetailsService;
+
+    beforeEach(module('hrjc'));
+
+    beforeEach(inject(function(_$controller_, _$rootScope_, _$httpBackend_, _$q_, _ContractService_,
+      _ContractDetailsService_) {
+      $controller = _$controller_;
+      $rootScope = _$rootScope_;
+      $httpBackend = _$httpBackend_;
+      ContractDetailsService = _ContractDetailsService_;
+      $q = _$q_;
+      entity = {
+        contract: {
+          id: '1',
+          contact_id: "04",
+          deleted: "0",
+          is_current: "1",
+          is_primary: "1"
+        },
+        details: {
+          id: "60",
+          position: "Test-added",
+          title: "Test-added",
+          funding_notes: null,
+          contract_type: "Apprentice",
+          period_start_date: "2017-03-28",
+          period_end_date: null,
+          end_reason: "1",
+          notice_amount: null,
+          notice_unit: null,
+          notice_amount_employee: null,
+          notice_unit_employee: null,
+          location: "Headquarters",
+          jobcontract_revision_id: "60"
+        },
+        hour: {},
+        pay: {},
+        leave: [],
+        health: {},
+        pension: {}
+      };
+
+      // $uibModalInstanceSpy for $uibModalInstance.opened.then()
+      $uibModalInstanceSpy = {
+        opened: {
+          then: function() {
+            jasmine.createSpy('modalInstance.opened');
+          }
+        }
+      };
+
+      // $uibModalSpy for $modal.open()
+      $uibModalSpy = {
+        open: function() {
+          return {
+            result: {
+              then: function(callback) {
+                callback("edit");
+              }
+            }
+          }
+        }
+      };
+
+      $httpBackend.whenGET(/action=validatedates&entity=HRJobDetails/).respond({ success: true });
+      $httpBackend.whenGET(/action=get&entity=HRJobContract/).respond({});
+      $httpBackend.whenPOST(/action=create&entity/).respond({});
+      $httpBackend.whenPOST(/action=replace&entity/).respond({});
+      $httpBackend.whenGET(/views.*/).respond({});
+
+      makeController();
+    }));
+
+
+    describe("Sets contract end date as appropriate", function() {
+
+      beforeEach(function() {
+        createModalSpy();
+        $scope.save();
+        $rootScope.$digest();
+      });
+
+      describe("When period_end_date is null", function() {
+
+        it("Sets contract period_end_date to '' ", function() {
+          expect($scope.entity.details.period_end_date).toBe("");
+        });
+      });
+
+      describe("When period_end_date is not null", function() {
```

Comment:
here is how this test case should be approached,

beforeEach(function() {
  1. Mock `ContractDetailsService.validateDates` so that the `then` is called, 
    confirm the `then` function is getting called using console.log or using debugger.
  2. Make result = true
  3. Make action = 'edit'
  4. Make `scope.entity.contract.is_primary == entity.contract.is_primary`
  5. set `$scope.entity.details.period_end_date` to null or a specific date, based on the test case
  6. Call the save function
  7. Any other steps required to resolve promises
  //All the above steps should be done in the describe block meant for save function only. They should be done in global describe block 
});

it('test', function () {
  actual test
  //MAke sure, that if you change the value here the test should fail.
})